Validate role templates before saving role mapping (#52636)
elastic/elasticsearch: x-pack/docs/en/rest-api/security/create-role-mappings.asciidoc

For more information, see <<mapping-roles>>.

> [[security-api-put-role-mapping-desc-templates]]

Reviewer: It's a bad practice to link to the auto-generated anchors, so it's unfortunate that one in the Kibana docs snuck through.  I think by explicitly defining that anchor, however, we should avoid breaking the link:
```suggestion
[[_role_templates]]
```

Author: Thanks Lisa. This is a clever way to solve the issue.

`rules`::
(Required, object) The rules that determine which users should be matched by the
mapping. A rule is a logical condition that is expressed by using a JSON DSL.
See  <<role-mapping-resources>>.

==== Role Templates

> NOTE: Role templates require the relevant scripting feature to be enabled,

Reviewer: I think this NOTE should move below the introductory paragraph, since at this point we haven't defined role templates yet.

Author: Thanks Lisa. Please make changes as you see fit. Thanks a lot.

Reviewer: Done, thanks!

> NOTE: Role templates require the relevant scripting feature to be enabled,

Reviewer: ```suggestion
NOTE: To use role templates successfully, the relevant scripting feature must be enabled.
```

> otherwise the creation of a role mapping with role templates will fail.

Reviewer: ```suggestion
Otherwise, all attempts to create a role mapping with role templates fail.
```

> <<allowed-script-types-setting>>.

Reviewer 2: We never call out what scripting features are related to which types of templates so "corresponding" might be confusing here. What's more,  we never call out the fact that you can reference a stored script in a template by id in any of the examples. Maybe something like: 

```suggestion
NOTE: Role templates require the relevant scripting feature to be enabled, otherwise the creation of a role mapping with role templates will fail. See <<allowed-script-types-setting>>.
```

and add an example where we reference a stored script template by id ?

<2> Because the template produces a JSON array, the format must be
    set to `json`.

> In addition to inline script, we can also use stored script for the

Reviewer: ```suggestion
In addition to an inline script, you can also use a stored script for the
```

`template` field. Above role mapping can also be created with the follows:

[source,console]
------------------------------------------------------------
POST /_scripts/derive-roles
{
  "script": {
      "lang": "mustache",
      "source" : "{{#tojson}}groups{{/tojson}}"
  }
}
------------------------------------------------------------

[source,console]
------------------------------------------------------------
POST /_security/role_mapping/mapping5
{
  "role_templates": [
    {
      "template": { "id": "derive-roles" },
      "format" : "json"
    }
  ],
  "rules": {
    "field" : { "realm.name" : "saml1" }
  },
  "enabled": true
}
------------------------------------------------------------
// TEST[continued]

> 

Reviewer 3: Can we think carefully about whether we really want to document & support this?
What's the use case? Why do we want to elevate this to "officially supported"?

Author: hmm good question. One issue of using stored script is consistency, i.e. the script can change without the role-mapping being aware of it. This could potentially creates confusion and support time frustration.  Is this the reason why we may not want to "officially" support it? (I assume adding it to official docs means offical support).

Reviewer 2: > What's the use case? Why do we want to elevate this to "officially supported"?

Can I flip the question ? Why do we "unofficially' support this already ? Why not disallow it instead ?

Author: I undersetand there are cases where we intentionally left them as a grey area, i.e. "unofficially supported". But I feel this is not the case here because:
1. Stored script support is _not_ an unique feature for this API. It is available and documented [elsewhere](https://www.elastic.co/guide/en/elasticsearch/reference/current/search-template.html#pre-registered-templates). So in some sense, it is already officially supported. There could be issues with stored scripts in general, which should be a separate discussion.
2. If we decide to disable the support here, it will be a breaking change even it is "unofficially" supported. So keeping it "unofficial" does not really buy us anything from this perspective.
3. General gain of clarity when something is clearly documented

So overall, I'd prefer to have it documented. Thanks

Reviewer 3: I'm not convinced.
What reason do we have for people to do this? Why should we spend time and effort maintaining the docs for this, when it has no clear value to anyone?

Author: Is there some context that I am not aware of, e.g. any past discussions about not having this documented in the first place? I admit that I don't have a strong reason other than "completeness" to add this in. But are there strong reasons to _not_ add it? I can think of a few candidates:

1. We don't want people to use stored script for authentication related stuff since a missing script is more devastating
2. We don't encourage stored script usage in general
3. We may want disable stored script for role template in future and by not documenting it we could avoid a breaking change
4. No user requires it 
5. We don't spend the extra time and effort because of 4

I'd be happy to go with item 1, 2, 3, but not particularly convinced by 4 and 5. I cannot say user would see value in this, but cannot say they would not either, unless there are past evidents of this being no value. Also the time and effort for this additional docs are not really significant. 

I am not really attached to this snippet of docs. I am happy to either remove or keep it. Just like to clear my thought process. I'd appreicate if you could elaborate it a bit.

Reviewer 3: Less is more.  
There's a temptation toward more docs (and lots of places where our docs are too brief) but it isn't always better.

One of our recurring experiences is that people try and configure ES security based on blogs because they seem simpler. The official docs have "more steps" so they follow a blog and then get stuck because those steps were actually necessarily and the blog skipped over them (or made assumptions that allowed them to ignore the steps) to keep it short.

Documenting something, tells people "this might be helpful to you" and also "you should read this if you want to understand how to use this feature". But I don't see any argument for _why_ either of those are true here. So what's the reason to do it?

Or, put another way, if a customer raises a support ticket saying "I'm trying to use a role mapping with a stored script as the template" my immediate reaction would be "Why on earth would you do that? That's a terrible idea!"  
And their (entirely reasonable) answer will be _because you documented it_.

> I cannot say user would see value in this

Then don't do it.
It's easy - if we cannot work out a reason why users would want to do it, then don't recommend it to them. Don't waste time writing and maintaining the docs. Don't waste their time by asking them to read those docs.

Author: > if a customer raises a support ticket saying "I'm trying to use a role mapping with a stored script as the template" my immediate reaction would be "Why on earth would you do that? That's a terrible idea!"

I can resonate with this argument. I do agree it's not a good idea to make a role template depend on external stuff. Since we cannot or don't want to recommend this usage ourselves, it makes sense to leave it out. Will update accordingly. Thanks

Reviewer 4: For what it’s worth, the role mapping UI in Kibana attempts to support the entire API surface. In other words, any valid role mapping created via the API should render and be editable in the UI. We of course default to inline scripts for these templates in the interface, but since stored scripts are technically available, we have to know how to render those as well, so there are UI controls to edit them.

I personally agree that these stored templates don’t make a lot of sense. Even if they aren’t documented, they’re more discoverable than they used to be because we now have a dedicated UI. For some folks, the UI will become the documentation, so like it or not, I fear we end up (currently) with the appearance that this is a first-class feature of role mappings.

We can consider changes to the UI which only surface these controls when editing an exiting mapping which already contains a stored script, but not without additional complexity

Author: Thanks for the insights @legrego 
You are right that the UI could be considered as documentation or even source of truth for some users. The UI also has a link to the doco ... it is possible that users would notice the difference and ask "why is there no document for this feature". 

For now, we are trying to not actively promote it. I wonder whether we could afford to disable it in 8.0. But I suspect that the solution could end up being adding more machinery to make this feature more robust, e.g. reference counting a stored script.
